cli/mender-artifact/write.go

```diff
@@ -333,6 +334,41 @@ func makeTypeInfo(ctx *cli.Context) (*artifact.TypeInfoV3, *artifact.TypeInfoV3,
 	return typeInfoV3, augmentTypeInfoV3, nil
 }
 
+func getSoftwareVersion(artifactName, softwareFilesystem, softwareName, softwareVersion string, noDefaultSoftwareVersion bool) map[string]string {
+	result := map[string]string{}
+	softwareVersionName := "rootfs-image"
+	if softwareFilesystem != "" {
+		softwareVersionName = softwareFilesystem
+	}
+	if softwareName != "" {
+		softwareVersionName += fmt.Sprintf(".%s", softwareName)
+	}
+	if noDefaultSoftwareVersion == false && softwareVersion == "" {
+		softwareVersion = artifactName
+	}
+	if softwareVersionName != "" && softwareVersion != "" {
+		result[softwareVersionName+".version"] = softwareVersion
+	}
+	return result
+}
+
+func applySoftwareVersionToTypeInfoProvides(ctx *cli.Context, typeInfoProvides artifact.TypeInfoProvides) artifact.TypeInfoProvides {
+	artifactName := ctx.String("artifact-name")
+	softwareFilesystem := ctx.String(softwareFilesystemFlag)
+	softwareName := ctx.String(softwareNameFlag)
+	softwareVersion := ctx.String(softwareVersionFlag)
+	noDefaultSoftwareVersion := ctx.Bool(noDefaultSoftwareVersionFlag)
+	if softwareVersionMapping := getSoftwareVersion(artifactName, softwareFilesystem, softwareName, softwareVersion, noDefaultSoftwareVersion); len(softwareVersionMapping) > 0 {
+		if typeInfoProvides == nil {
+			typeInfoProvides = make(map[string]string)
+		}
+		for key, value := range softwareVersionMapping {
+			typeInfoProvides[key] = value
+		}
```

Comment:
Since this is not clear from the function signature, I think there should be a comment on this function: "may modify the original map, or return a new one".

Reply:
I updated the function to NOT modify the mapping passed in as argument, and added a comment.